enhancements/testing/improved-platform-tests.md

```diff
@@ -0,0 +1,412 @@
+<!--
+Above diagram was generated at https://mermaid.live/ using following code:
+sequenceDiagram
+    actor U as User
+    participant O as openshift-tests
+    participant E as external-tests
+    participant R as release:tag
+    U->>+O: run openshift/conformance
+    O->>+O: list tests
+    O->>+R: extract test binary
+    rect rgb(177, 177, 177)
+        opt test binary exists?
+            R->>+E: external-tests
+            O->>+E: list tests
+            E->>+O: tests
+        end
+    end
+    loop over all tests
+        alt embedded test?
+            O->>+O: run #lt;test name#gt;#10;
+        else external test?
+            rect rgb(177, 177, 177)
+                O->>+E: run #lt;test name#gt;
+                E->>+O: results
+            end
+        end
+    end
+    O->>+U: results
+-->
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints
+
+To achieve the first step of the proposal, we need to:
+1. Provide a library for building `<binary>-tests` which exposes two commands:
+   - `list` - responsible for listing tests in JSON format;
+   - `run` - run a single test, returning results in ginkgo compatible format;
+2. Provide an image annotation or similar mechanism informing where the test
+   binary is embedded in the release payload.
+
+The proposed extensions to `openshift-tests` will require the following changes:
+1. Add extraction logic responsible for retrieving test binaries from the
+   release payload, based on the annotations provided by the authors of these
+   images.
+2. Add logic responsible for iterating over all test binaries invoking the `list`
+   subcommand, and reading the list of tests provided by the binary. Every tests
+   should have an associated binary name, which will be used during invocation.
+3. Expand the current filtering, shuffling, and randomization logic to also take
+   into account the new tests coming from external binaries.
+4. Add logic responsible for iterating over all tests, invoking the `run`
+   subcommand using the associated test's binary and reading its results.
+5. Provide a fallback path allowing to use the vendored tests, only when either
+   the release image is not available, or on demand (`OPENSHIFT_SKIP_EXTERNAL_TESTS`
+   environment variable, for example).
+
+### Risks and Mitigations
+
+- The current policy when the [origin](https://github.com/openshift/origin/) pull
+  request bringing the updated k8s version breaks the CI is to revert PR. This
+  policy will be harder to execute with this change because reverting the o/k
+  rebase PR can be problematic and was never tried before.
+
+There should be fewer reasons to do any reverting, because with this new mechanism
+we should get the appropriate signal during extensive payload testing required
+to land every k8s bump PR.
+
+- Some of the work that has been done to build the partner certification tool will
+  need to be updated to pull additional images with separate test binaries.
```

Comment:
this seems to point to a design aspect that isn't clearly specified yet.

Is the intent to have each component/externalized test binary go into its own image?

Or is the intent to bring all the externalized test binaries into the existing `tests` image? 

I prefer the latter, though it's more effort at build time...but it's less effort for teams to adopt (they don't need to start publishing a new image), doesn't grow the payload size as much, and avoids the problem of not being able pull certain images.

it also makes the list of test binaries obvious:  the `tests` image could contain a list of binaries, or even just a well known directory containing all the binaries.  If we maintain them in separate images, then something needs to make it clear what the list of component-test images is.

Reply:
The goal of this specific point is to ensure folks building on top of `openshift-tests` that all the additional work will be handled within binary and no additional work should be required by consumers.

Reply:
> The goal of this specific point is to ensure folks building on top of openshift-tests that all the additional work will be handled within binary and no additional work should be required by consumers.

in that case i think you should just state that assurance generically.  As it's currently written it assumes a specific implementation (image per test binary).

Reply:
Updated.